// react-native-rag/lib/module/text-splitters/langchain.js
"use strict";

import { CharacterTextSplitter as LangchainCharacterTextSplitter, RecursiveCharacterTextSplitter as LangchainRecursiveCharacterTextSplitter, TokenTextSplitter as LangchainTokenTextSplitter, MarkdownTextSplitter as LangchainMarkdownTextSplitter, LatexTextSplitter as LangchainLatexTextSplitter } from '@langchain/textsplitters';

/**
 * Common parameters for configuring text splitters.
 */

/**
 * A text splitter that splits text based on a fixed character count.
 * This is a wrapper around Langchain's `CharacterTextSplitter`.
 */
export class CharacterTextSplitter {
  /**
   * Creates an instance of CharacterTextSplitter.
   * @param params - Parameters for the text splitter.
   * @param params.chunkSize - The maximum size of each text chunk.
   * @param params.chunkOverlap - The amount of overlap between consecutive text chunks.
   */
  constructor({
    chunkSize,
    chunkOverlap
  }) {
    this.textSplitter = new LangchainCharacterTextSplitter({
      chunkSize,
      chunkOverlap
    });
  }

  /**
   * Splits the given text into chunks based on character count.
   * @param text - The input text to be split.
   * @returns A promise that resolves to an array of string chunks.
   */
  async splitText(text) {
    return this.textSplitter.splitText(text);
  }
}

/**
 * A text splitter that attempts to preserve semantic boundaries by
 * recursively trying different separators.
 * This is a wrapper around Langchain's `RecursiveCharacterTextSplitter`.
 */
export class RecursiveCharacterTextSplitter {
  /**
   * Creates an instance of RecursiveCharacterTextSplitter.
   * @param params - Parameters for the text splitter.
   * @param params.chunkSize - The maximum size of each text chunk.
   * @param params.chunkOverlap - The amount of overlap between consecutive text chunks.
   */
  constructor({
    chunkSize,
    chunkOverlap
  }) {
    this.textSplitter = new LangchainRecursiveCharacterTextSplitter({
      chunkSize,
      chunkOverlap
    });
  }

  /**
   * Splits the given text into chunks using a recursive character splitting strategy.
   * @param text - The input text to be split.
   * @returns A promise that resolves to an array of string chunks.
   */
  async splitText(text) {
    return this.textSplitter.splitText(text);
  }
}

/**
 * A text splitter that splits text based on a token count rather than character count.
 * This is a wrapper around Langchain's `TokenTextSplitter`.
 */
export class TokenTextSplitter {
  /**
   * Creates an instance of TokenTextSplitter.
   * @param params - Parameters for the text splitter.
   * @param params.chunkSize - The maximum size of each text chunk (in tokens).
   * @param params.chunkOverlap - The amount of overlap between consecutive text chunks (in tokens).
   */
  constructor({
    chunkSize,
    chunkOverlap
  }) {
    this.textSplitter = new LangchainTokenTextSplitter({
      chunkSize,
      chunkOverlap
    });
  }

  /**
   * Splits the given text into chunks based on token count.
   * @param text - The input text to be split.
   * @returns A promise that resolves to an array of string chunks.
   */
  async splitText(text) {
    return this.textSplitter.splitText(text);
  }
}

/**
 * A text splitter specifically designed for Markdown documents,
 * attempting to preserve Markdown structure within chunks.
 * This is a wrapper around Langchain's `MarkdownTextSplitter`.
 */
export class MarkdownTextSplitter {
  /**
   * Creates an instance of MarkdownTextSplitter.
   * @param params - Parameters for the text splitter.
   * @param params.chunkSize - The maximum size of each text chunk.
   * @param params.chunkOverlap - The amount of overlap between consecutive text chunks.
   */
  constructor({
    chunkSize,
    chunkOverlap
  }) {
    this.textSplitter = new LangchainMarkdownTextSplitter({
      chunkSize,
      chunkOverlap
    });
  }

  /**
   * Splits the given Markdown text into chunks.
   * @param text - The input Markdown text to be split.
   * @returns A promise that resolves to an array of string chunks.
   */
  async splitText(text) {
    return this.textSplitter.splitText(text);
  }
}

/**
 * A text splitter specifically designed for LaTeX documents,
 * attempting to preserve LaTeX structure within chunks.
 * This is a wrapper around Langchain's `LatexTextSplitter`.
 */
export class LatexTextSplitter {
  /**
   * Creates an instance of LatexTextSplitter.
   * @param params - Parameters for the text splitter.
   * @param params.chunkSize - The maximum size of each text chunk.
   * @param params.chunkOverlap - The amount of overlap between consecutive text chunks.
   */
  constructor({
    chunkSize,
    chunkOverlap
  }) {
    this.textSplitter = new LangchainLatexTextSplitter({
      chunkSize,
      chunkOverlap
    });
  }

  /**
   * Splits the given LaTeX text into chunks.
   * @param text - The input LaTeX text to be split.
   * @returns A promise that resolves to an array of string chunks.
   */
  async splitText(text) {
    return this.textSplitter.splitText(text);
  }
}
