// react-native-executorch/lib/module/types/common.js
"use strict";

export let ScalarType = /*#__PURE__*/function (ScalarType) {
  ScalarType[ScalarType["BYTE"] = 0] = "BYTE";
  ScalarType[ScalarType["CHAR"] = 1] = "CHAR";
  ScalarType[ScalarType["SHORT"] = 2] = "SHORT";
  ScalarType[ScalarType["INT"] = 3] = "INT";
  ScalarType[ScalarType["LONG"] = 4] = "LONG";
  ScalarType[ScalarType["HALF"] = 5] = "HALF";
  ScalarType[ScalarType["FLOAT"] = 6] = "FLOAT";
  ScalarType[ScalarType["DOUBLE"] = 7] = "DOUBLE";
  ScalarType[ScalarType["BOOL"] = 11] = "BOOL";
  ScalarType[ScalarType["QINT8"] = 12] = "QINT8";
  ScalarType[ScalarType["QUINT8"] = 13] = "QUINT8";
  ScalarType[ScalarType["QINT32"] = 14] = "QINT32";
  ScalarType[ScalarType["QUINT4X2"] = 16] = "QUINT4X2";
  ScalarType[ScalarType["QUINT2X4"] = 17] = "QUINT2X4";
  ScalarType[ScalarType["BITS16"] = 22] = "BITS16";
  ScalarType[ScalarType["FLOAT8E5M2"] = 23] = "FLOAT8E5M2";
  ScalarType[ScalarType["FLOAT8E4M3FN"] = 24] = "FLOAT8E4M3FN";
  ScalarType[ScalarType["FLOAT8E5M2FNUZ"] = 25] = "FLOAT8E5M2FNUZ";
  ScalarType[ScalarType["FLOAT8E4M3FNUZ"] = 26] = "FLOAT8E4M3FNUZ";
  ScalarType[ScalarType["UINT16"] = 27] = "UINT16";
  ScalarType[ScalarType["UINT32"] = 28] = "UINT32";
  ScalarType[ScalarType["UINT64"] = 29] = "UINT64";
  return ScalarType;
}({});
