// react-native-executorch/lib/module/types/imageSegmentation.js
"use strict";

/* eslint-disable @cspell/spellchecker */
export let DeeplabLabel = /*#__PURE__*/function (DeeplabLabel) {
  DeeplabLabel[DeeplabLabel["BACKGROUND"] = 0] = "BACKGROUND";
  DeeplabLabel[DeeplabLabel["AEROPLANE"] = 1] = "AEROPLANE";
  DeeplabLabel[DeeplabLabel["BICYCLE"] = 2] = "BICYCLE";
  DeeplabLabel[DeeplabLabel["BIRD"] = 3] = "BIRD";
  DeeplabLabel[DeeplabLabel["BOAT"] = 4] = "BOAT";
  DeeplabLabel[DeeplabLabel["BOTTLE"] = 5] = "BOTTLE";
  DeeplabLabel[DeeplabLabel["BUS"] = 6] = "BUS";
  DeeplabLabel[DeeplabLabel["CAR"] = 7] = "CAR";
  DeeplabLabel[DeeplabLabel["CAT"] = 8] = "CAT";
  DeeplabLabel[DeeplabLabel["CHAIR"] = 9] = "CHAIR";
  DeeplabLabel[DeeplabLabel["COW"] = 10] = "COW";
  DeeplabLabel[DeeplabLabel["DININGTABLE"] = 11] = "DININGTABLE";
  DeeplabLabel[DeeplabLabel["DOG"] = 12] = "DOG";
  DeeplabLabel[DeeplabLabel["HORSE"] = 13] = "HORSE";
  DeeplabLabel[DeeplabLabel["MOTORBIKE"] = 14] = "MOTORBIKE";
  DeeplabLabel[DeeplabLabel["PERSON"] = 15] = "PERSON";
  DeeplabLabel[DeeplabLabel["POTTEDPLANT"] = 16] = "POTTEDPLANT";
  DeeplabLabel[DeeplabLabel["SHEEP"] = 17] = "SHEEP";
  DeeplabLabel[DeeplabLabel["SOFA"] = 18] = "SOFA";
  DeeplabLabel[DeeplabLabel["TRAIN"] = 19] = "TRAIN";
  DeeplabLabel[DeeplabLabel["TVMONITOR"] = 20] = "TVMONITOR";
  DeeplabLabel[DeeplabLabel["ARGMAX"] = 21] = "ARGMAX"; // Additional label not present in the model
  return DeeplabLabel;
}({});
