// react-native-executorch/lib/module/hooks/computer_vision/useVerticalOCR.js
"use strict";

import { useEffect, useMemo, useState } from 'react';
import { VerticalOCRController } from '../../controllers/VerticalOCRController';
export const useVerticalOCR = ({
  model,
  independentCharacters = false,
  preventLoad = false
}) => {
  const [error, setError] = useState(null);
  const [isReady, setIsReady] = useState(false);
  const [isGenerating, setIsGenerating] = useState(false);
  const [downloadProgress, setDownloadProgress] = useState(0);
  const controllerInstance = useMemo(() => new VerticalOCRController({
    isReadyCallback: setIsReady,
    isGeneratingCallback: setIsGenerating,
    errorCallback: setError
  }), []);
  useEffect(() => {
    if (preventLoad) return;
    (async () => {
      await controllerInstance.load({
        detectorLarge: model.detectorLarge,
        detectorNarrow: model.detectorNarrow
      }, {
        recognizerLarge: model.recognizerLarge,
        recognizerSmall: model.recognizerSmall
      }, model.language, independentCharacters, setDownloadProgress);
    })();
    return () => {
      controllerInstance.delete();
    };
  }, [controllerInstance, model.detectorLarge, model.detectorNarrow, model.recognizerLarge, model.recognizerSmall, model.language, independentCharacters, preventLoad]);
  return {
    error,
    isReady,
    isGenerating,
    forward: controllerInstance.forward,
    downloadProgress
  };
};
